File: frontend/src/app/core/services/project.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { ApiResponse, Member, Project, ProjectRequest, ProjectRole } from '../models';

@Injectable({ providedIn: 'root' })
export class ProjectService {
  private readonly apiUrl = `${environment.apiUrl}/projects`;

  constructor(private http: HttpClient) {}

  getProjects(): Observable<ApiResponse<Project[]>> {
    return this.http.get<ApiResponse<Project[]>>(this.apiUrl);
  }

  getProject(id: number): Observable<ApiResponse<Project>> {
    return this.http.get<ApiResponse<Project>>(`${this.apiUrl}/${id}`);
  }

  createProject(request: ProjectRequest): Observable<ApiResponse<Project>> {
    return this.http.post<ApiResponse<Project>>(this.apiUrl, request);
  }

  updateProject(id: number, request: ProjectRequest): Observable<ApiResponse<Project>> {
    return this.http.put<ApiResponse<Project>>(`${this.apiUrl}/${id}`, request);
  }

  deleteProject(id: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${id}`);
  }

  getMembers(projectId: number): Observable<ApiResponse<Member[]>> {
    return this.http.get<ApiResponse<Member[]>>(`${this.apiUrl}/${projectId}/members`);
  }

  addMember(projectId: number, userId: number, role: ProjectRole = 'MEMBER'): Observable<ApiResponse<Member>> {
    return this.http.post<ApiResponse<Member>>(`${this.apiUrl}/${projectId}/members`, { userId, role });
  }

  removeMember(projectId: number, memberId: number): Observable<ApiResponse<void>> {
    return this.http.delete<ApiResponse<void>>(`${this.apiUrl}/${projectId}/members/${memberId}`);
  }
}

File: backend/src/main/java/com/cirquetask/service/ProjectService.java
package com.cirquetask.service;

import com.cirquetask.model.dto.MemberDto;
import com.cirquetask.model.dto.ProjectDto;
import com.cirquetask.model.dto.ProjectRequest;
import com.cirquetask.model.enums.ProjectRole;

import java.util.List;

public interface ProjectService {

    ProjectDto createProject(ProjectRequest request, Long userId);

    ProjectDto getProject(Long projectId, Long userId);

    List<ProjectDto> getUserProjects(Long userId);

    ProjectDto updateProject(Long projectId, ProjectRequest request, Long userId);

    void deleteProject(Long projectId, Long userId);

    MemberDto addMember(Long projectId, Long memberUserId, ProjectRole role, Long currentUserId);

    void removeMember(Long projectId, Long memberUserId, Long currentUserId);

    MemberDto updateMemberRole(Long projectId, Long memberUserId, ProjectRole role, Long currentUserId);

    List<MemberDto> getMembers(Long projectId, Long userId);
}
